src/sparseml/pytorch/sparsification/distillation/modifier_per_layer.py

```diff
@@ -0,0 +1,442 @@
+    BaseDistillationModifier,
+)
+from sparseml.pytorch.sparsification.modifier import PyTorchModifierYAML
+from sparseml.pytorch.utils import BaseLogger
+
+
+__all__ = [
+    "PerLayerDistillationModifier",
+]
+
+_LOGGER = logging.getLogger(__name__)
+
+_DISTILLATION_TYPES = [torch.nn.Conv2d, torch.nn.Linear]
+
+
+@PyTorchModifierYAML()
+class PerLayerDistillationModifier(BaseDistillationModifier):
+    """
+    Adds a knowledge distillation loss based on the feature imitation loss.
+    A distillation_teacher module may be provided as a kwarg to
+    the Manager initialization and loss_update(loss) must be called before any
+    backwards pass in the integrated training flow.
+
+    If no teacher model is provided, then self-distillation will be used.
+    The feature difference between teacher and student can be weighted spatially
+    by a weighing function.
+
+    The connection between layers in the student model and layers in the teacher
+    model is controlled via the `teacher_layer_names` and `student_layer_names`
+    fields. **These fields must be the same length!** The i'th item in both
+    of these arrays are paired together.
+
+    # Sample yaml:
+
+    ```yaml
+    !PerLayerDistillationModifier
+        gain: 2.0
+        start_epoch: 0.0
+        end_epoch: 1.0
+        update_frequency: 0.2
+        normalize: True
+        student_layer_names: ["layer2.0.conv2"]
+        teacher_layer_names: ["layer2.0.conv2"]
+        project_features: True
+        epsilon: 1e-6
+    ```
+
+    # Parameters
+
+    :param gain: How much to weight the distillation loss.
+        Default is `1.5`
+    :param start_epoch: The epoch to start the modifier at
+    :param end_epoch: The epoch to end the modifier at
+    :param update_frequency: The number of epochs or fraction of epochs to
+            update at between start and end
+    :param normalize: Whether to normalize the output difference by the
+        the magnitude of the teacher's output.
+        Default is `True`.
+    :param student_layer_names: List of layer names to distill.
+        *Must be same length as teacher_layer_names.*
+    :param teacher_layer_names: List of layer names to distill from.
+        *Must be same length as student_layer_names.*
+    :param project_features: Whether to project the output of student layers to the
+        same size as the output of the teacher layers.
+        Default is `True`
+    :param epsilon: Small value used to avoid division by zero when normalization
+        is used. Default is `1e-6`
+    """
+
+    def __init__(
+        self,
+        gain: float = 1.5,
+        start_epoch: float = -1.0,
+        end_epoch: float = -1.0,
+        update_frequency: float = -1.0,
+        normalize: bool = True,
+        student_layer_names: Optional[List[str]] = None,
+        teacher_layer_names: Optional[List[str]] = None,
+        project_features: bool = True,
+        epsilon: float = 1e-6,
+    ):
+        if (
```

Comment:
I think that is technically allowed since in that case both would just default to all layers (and then there's a value error raised in initialize if they have different number of layers)